### ISSUE TITLE
Clicking the "add labels" button when viewing a lazy (zarr/dask) image should create a zarr-backed labels array.

### ISSUE THREAD

**Issue by jni** [napari/napari#470]
## 🚀 Feature

Currently, add_labels will try to create an in-memory labels array of the same shape as the currently viewed image. This will break if the image layer is a bigger-than-memory dask array. A possible solution is to create an out-of-core dataset backed by zarr.

**Comment by sofroniewn** [comment]
@manzt do you have any thoughts on if you could make a zarr "plugin" for labels painting? maybe in the first case you'd want to just explore the feature directly before seeing if it could be funneled through our current plugin API.

**Comment by manzt** [comment]
I think I would need to take a closer look at the hook specification (since I've only implemented `napari_get_reader`), but zarr would be a very efficient way to back the labels array, especially if labels are sparse.

> Currently, add_labels will try to create an in-memory labels array of the same shape as the currently viewed image.

You could by default create an in-memory `zarr.Array` which is very light-weight:

```python
>>> import zarr
>>> labels = zarr.zeros(image.shape, dtype=??) # let zarr determine chunks
>>> labels
<zarr.core.Array (10000, 10000) int32
>>> isinstance(labels.store, dict)
True
>>> labels.store.keys()
dict_keys(['.zarray'])
```

By design, `zarr.zeros` is simply a creation method where only metadata for the array is saved to the underlying store, in this case a python `dict`. Chunks are only created on write, and if you try to access a missing chunks, zarr will fill using the `fill_value` (`0`) on read. 

```python
>>> labels[:]
array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int32)
```

This means you could create a label array with a massive shape and only run into issues with memory when you start to write lots of chunks. Fortunately with zarr, you can actually keep track of how many chunks have been initialized as well as the number of bytes currently stored. 

```python
>>> labels.info
Type               : zarr.core.Array
Data type          : int32
Shape              : (10000, 10000)
Chunk shape        : (625, 625)
Order              : C
Read-only          : False
Compressor         : Blosc(cname='lz4', clevel=5, shuffle=SHUFFLE, blocksize=0)
Store type         : builtins.dict
No. bytes          : 400000000 (381.5M)
No. bytes stored   : 343
Storage ratio      : 1166180.8
Chunks initialized : 0/256
```

With this information, you could decide whether to move the in-memory store (`dict`) to a `zarr.DirectoryStore` and make the labels out-of-core only when it starts to get large. 

EDIT: I could even imagine a GUI prompt: "The 'labels' array is starting to get large, would you like to move the store to a directory?"

**Comment by sofroniewn** [comment]
> You could by default create an in-memory zarr.Array which is very light-weight:

> you could decide whether to move the in-memory store (dict) to a zarr.DirectoryStore and make the labels out-of-core only when it starts to get large.

> I could even imagine a GUI prompt: "The 'labels' array is starting to get large, would you like to move the store to a directory?"

Wow i love these ideas, curious what @jni thinks, but you may have broken this one wide open :-) 

I think I'd probably look at the total number of pixels requested by the new labels layer, if it was below a threshold I'd just use a numpy array (as people might want an easy way to save labels to a tiff), and if it was larger then i'd start with the in memory zarr array, and then moving to the zarr directory store sounds like a great idea.

in general we need to think more about how to do the selection/ little context menus / GUI prompts

**Comment by manzt** [comment]
Totally agree! numpy by default (if below a certain pixel threshold), and only zarr if it makes sense (with flexible storage backends).

**Comment by manzt** [comment]
Putting this here for future reference. Dug into the Zarr docs and found they have a convenience method for coping one store to the other (in-memory to directory in this case).

```python
import zarr
labels = zarr.zeros((10_000, 10_000))
dir_store = zarr.DirectoryStore('labels.zarr') 
zarr.copy_store(labels.store, dir_store) # copies current store to some other backend
```

**Comment by kephale** [comment]
@jni @DragaDoncila is this still relevant, or does https://github.com/jni/zarpaint address the need?

**Comment by jni** [comment]
I think it's still relevant — users right now can shoot themselves in the foot very easily with that button, so I think the default (no plugins) behaviour is important.

Importantly, zarr arrays are lazy by default (no data is created if chunks are empty) *and* labels compress extremely well, so I think a "close enough" solution here would be to use an in-memory zarr store. This would still be "theoretically" huge but in practice it would be ~zero cost to create and very cheap to use.

**Comment by psobolewskiPhD** [comment]
Bumping this, I've run into this and have had users report it when trying napari with whole-slide-images (WSI), yielding "napari really sucks!"

**Comment by Czaki** [comment]
Does it mean that it should open file dialog to select place to save zarr array?

**Comment by jni** [comment]
I'm thinking let's start with in-memory zarr. Then we can write to zarr on-disk using a writer plugin. In the future, clicking on the labels button might open a three item menu — "make numpy array" "make in-memory zarr array" and "make on-disk zarr array" — this last one opening a file dialog.

But, in memory zarr is so cost-effective that I think we should do that straight away, and use it automatically when e.g. the array would be >500MB.

**Comment by Czaki** [comment]
> But, in memory zarr is so cost-effective that I think we should do that straight away, and use it automatically when e.g. the array would be >500MB.

So in some moment, there should be a file dialog popup? 



> I'm thinking let's start with in-memory zarr

As I understand the whole problem is a memory error, so I do not see how creating in memory zarr may solve this. Or Zarr support in-memory compression?

**Comment by manzt** [comment]
> As I understand the whole problem is a memory error, so I do not see how creating in memory zarr may solve this. Or Zarr support in-memory compression?

I like to think of Zarr as a configurable pipeline for chunking and compressing arrays. For in-memory zarr, the compressed chunks are just stored in a dict, which may be be accessed/updated on demand.

**Comment by jni** [comment]
Yes, zarr supports [in-memory compression](https://zarr.readthedocs.io/en/stable/api/storage.html#zarr.storage.MemoryStore) (indeed that was the [original use-case](https://twitter.com/jnuneziglesias/status/679230849275957249)), *and* zero-chunks are implicit, so creating a zarr array costs nothing, and adding data adds compressed data.

**Comment by jni** [comment]
ie @Czaki I don't think there should be a file dialog popup in the first implementation. Only an in-memory zarr. That avoids having to make *any* UI changes, which is nice.

**Comment by Czaki** [comment]
The #6190 is ready for review
